Persist currently started allocation IDs to index metadata

ywelsch opened the pull request:
These allocation IDs serve as candidates to decide future primary shards

Subtask of #14739

s1monw commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -218,6 +226,7 @@ private IndexMetaData(String index, long version, State state, Settings settings
         this.settings = settings;
         this.mappings = mappings;
         this.customs = customs;
+        this.primaryCandidateAllocationIds = primaryCandidateAllocationIds;
```
I think we should make this unmodifiable here at least?

s1monw commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -550,6 +575,7 @@ public void writeTo(StreamOutput out) throws IOException {
             out.writeString(cursor.key);
             cursor.value.writeTo(out);
         }
+        out.writeStringArray(primaryCandidateAllocationIds.toArray(new String[primaryCandidateAllocationIds.size()]));
```
I am not sure if the iteration order is important somewhere down the road but if we serialize over the network and build another set from it we can end up with different iteration order. If it's not important that's fine just wanna raise it.

bleskes replied:
I think that's OK, at least for now.  We only care about member of operations.

s1monw commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -714,7 +753,8 @@ public IndexMetaData build() {
 
-            return new IndexMetaData(index, version, state, tmpSettings, mappings.build(), tmpAliases.build(), customs.build());
+            return new IndexMetaData(index, version, state, tmpSettings, mappings.build(), tmpAliases.build(), customs.build(),
+                    Collections.unmodifiableSet(new HashSet<>(primaryCandidateAllocationIds)));
```
lets pass the original set here and make it unmodifiable in the ctor

s1monw commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -816,6 +861,12 @@ public static IndexMetaData fromXContent(XContentParser parser) throws IOExcepti
                                 }
                             }
                         }
+                    } else if ("primary_candidate_allocation_ids".equals(currentFieldName)) {
```
can this be a string constant?

s1monw commented:
left minor comments LGTM otherwise

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -194,7 +200,9 @@ public static State fromString(String state) {
-    private IndexMetaData(String index, long version, State state, Settings settings, ImmutableOpenMap<String, MappingMetaData> mappings, ImmutableOpenMap<String, AliasMetaData> aliases, ImmutableOpenMap<String, Custom> customs) {
+    private IndexMetaData(String index, long version, State state, Settings settings, ImmutableOpenMap<String, MappingMetaData> mappings,
+                          ImmutableOpenMap<String, AliasMetaData> aliases, ImmutableOpenMap<String, Custom> customs,
+                          Set<String> primaryCandidateAllocationIds) {
```
can we call these "activeAllocationIds" ?

bleskes replied:
also we need those *PER* shard.. so the allocation ids for shard 0, alloc ids for shard 1 etc.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -757,6 +797,11 @@ public static void toXContent(IndexMetaData indexMetaData, XContentBuilder build
             }
             builder.endObject();
 
+            builder.startArray("primary_candidate_allocation_ids");
```
"active_allocations" ? +1 on making it a constant.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -654,6 +683,16 @@ public Builder putMapping(MappingMetaData mappingMd) {
+            return this;
+        }
+
+        public Builder addPrimaryCandidateAllocationId(String allocationId) {
```
I think this will be simpler if we allow setting all the set in one go (see other comment in AllocationService)

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -79,24 +85,77 @@ public AllocationService(Settings settings, AllocationDeciders allocationDecider
+            }
+            IndexMetaData.Builder indexMetaDataBuilder = null;
+            for (IndexShardRoutingTable shardRoutings : indexRoutingTable) {
+                for (ShardRouting shardRouting : shardRoutings.activeShards()) {
```
I think this is easier to understand as it makes a 1-1 copy of the current active shard allocations in the routing table:

```
for (IndexShardRoutingTable shardRoutings : indexRoutingTable) {
    Set<AllocationId> activeShards = shardRoutings.activeShards().stream()
            .map(shardRouting -> shardRouting.allocationId())
            .filter(allocationId -> allocationId != null)
            .collect(Collectors.toSet());
    if (activeShards.isEmpty() == false && activeShards.equals(indexMetaData.getActiveShards(shardRoutings.shardId().id())) == false) {
        // only update active allocation ids if there is an active shard
        if (indexMetaDataBuilder == null) {
            indexMetaDataBuilder = IndexMetaData.builder(indexMetaData);
        }

        indexMetaDataBuilder.setActiveAllocations(shardRoutings.shardId().id(), activeShards);
    }
}
```

bleskes commented on core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -307,6 +307,13 @@ private void beginSnapshot(ClusterState clusterState, final SnapshotsInProgress.
+            // remove allocation ids from metadata
+            MetaData.Builder builder = MetaData.builder(metaData);
+            for (ObjectObjectCursor<String, IndexMetaData> index : metaData.getIndices()) {
+                builder.put(IndexMetaData.builder(index.value).removeAllPrimaryCandidateAllocationIds().build(), false);
```
this is a good one. Do you know what shard state we generate when we are restoring?

ywelsch replied:
Upon restore we know the RestoreSource. In that case, we can have special handling. I remove this for now and keep this in mind when implementing the actual check whether allocation ids match.

bleskes commented:
This looks good. I left some minor suggestion. Note the comment about storing allocation ids per shard in the meta data. I know the allocation Ids are unique so in theory we can throw them in one big pile, but I think it will be clearer in terms of code and API later on to have them separated.

ywelsch commented:
Pushed a larger set of changes, please have another look @bleskes and @s1monw. I store the allocation ids now on a per shard id basis and introduced a new class that helps with the serialization / cluster diffs.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -235,47 +302,88 @@ protected MapDiff() {
         }
     }
+
+    private static abstract class StringKeyedMapDiff<T extends Diffable<T>, M> extends MapDiff<String, T, M> {
```
I wonder if we should follow the OpenIntMap and call this StringMapDiff , just to key things shorter.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -235,47 +302,88 @@ protected MapDiff() {
+        }
+    }
+
+    private static abstract class IntKeyedMapDiff<T extends Diffable<T>, M> extends MapDiff<Integer, T, M> {
```
same suggestion -> IntMapDiff

bleskes commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -235,47 +302,88 @@ protected MapDiff() {
+
+    private static abstract class IntKeyedMapDiff<T extends Diffable<T>, M> extends MapDiff<Integer, T, M> {
+        protected IntKeyedMapDiff() {
+            super();
```
Those are implicit. Can be removed.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/AllocationIds.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.Set;
+
+/**
+ * Represents a set of allocation ids
```
Can we add a note that we need a set that implements Diffable?

bleskes replied:
I wonder if we should just make this a generic DiffableStringSet?

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -364,6 +375,14 @@ public MappingMetaData mappingOrDefault(String mappingType) {
         return (T) customs.get(type);
     }
 
+    public ImmutableOpenIntMap<AllocationIds> getActiveAllocationIds() {
```
do we need to expose this? I rather not and iterate using a shardId int.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -450,7 +473,8 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         private final Settings settings;
         private final Diff<ImmutableOpenMap<String, MappingMetaData>> mappings;
         private final Diff<ImmutableOpenMap<String, AliasMetaData>> aliases;
-        private Diff<ImmutableOpenMap<String, Custom>> customs;
```
good catch

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -218,6 +228,7 @@ private IndexMetaData(String index, long version, State state, Settings settings
         this.settings = settings;
         this.mappings = mappings;
         this.customs = customs;
+        this.activeAllocationIds = activeAllocationIds;
```
can we make sure we prepopulate this map and have an empty set for every shard?

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -693,6 +735,25 @@ public Custom getCustom(String type) {
+            return this;
+        }
+
+        public Builder removeActiveAllocationIds(int shardId) {
```
these do not seem to be used. I don't thing we need them? I rather make sure there is always a set for each shard.

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/ActiveAllocationIdTests.java:
```diff
@@ -0,0 +1,64 @@
+        clusterState = ClusterState.builder(clusterState).routingResult(rerouteResult).build();
+
+        assertThat(clusterState.metaData().index("test").activeAllocationIds(0), nullValue());
+        assertThat(clusterState.metaData().index("test-old").activeAllocationIds(0).asSet().size(), equalTo(2));
```
can we check that these are "x" & "y"?

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/ActiveAllocationIdTests.java:
```diff
@@ -0,0 +1,64 @@
+        rerouteResult = allocation.applyStartedShards(clusterState, clusterState.getRoutingNodes().shardsWithState(INITIALIZING));
+        clusterState = ClusterState.builder(clusterState).routingResult(rerouteResult).build();
+
+        assertThat(clusterState.metaData().index("test").activeAllocationIds(0).asSet().size(), equalTo(1));
```
can we check this one is equal to what we expect?

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/ActiveAllocationIdTests.java:
```diff
@@ -0,0 +1,64 @@
+        rerouteResult = allocation.reroute(clusterState, "reroute");
+        clusterState = ClusterState.builder(clusterState).routingResult(rerouteResult).build();
+
+        assertThat(clusterState.metaData().index("test").activeAllocationIds(0).asSet().size(), equalTo(2));
```
can we add a check that when we loose all copies we don't update the allocation id list?

bleskes commented:
thx @ywelsch . This looks great. Left some comments here and there. One thing I think is forgotten (because it's not evident at all) is adding the allocation ids to the toXContent serialization of the cluster state.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -693,6 +735,25 @@ public Custom getCustom(String type) {
+            return this;
+        }
+
+        public AllocationIds getActiveAllocationIds(int shardId) {
```
It would be great, if we move to a DiffableStringSet, to make this version return a Set<String> and bellow the hood return an immutable set.

ywelsch commented:
Pushed another set of changes. In IndexMetaData, the activeAllocationIds are now stored as `ImmutableOpenIntMap<Set<String>>`. To enable serialization for maps that have non-diffable values, I heavily refactored DiffableUtils. I also introduced the possibility to specify how key values of map should be serialized. For activeAllocationIds (map from shard id to set of allocation id), we can use VInt-based serialization of keys (as we know that shard ids are small positive numbers).

bleskes commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -19,263 +19,624 @@
+            }
+            for (Map.Entry<K, T> partIter : after.entrySet()) {
+                T beforePart = before.get(partIter.getKey());
+                if (beforePart == null || (valueSerializer.supportsDiffableValues() == false)) {
```
isn't the `valueSerializer.supportsDiffableValues() == false` only relevant if the before and after are not equal?

ywelsch replied:
you're absolutely right. will fix.

bleskes replied:
still busy review so I didn't review the unit tests yet - but this does seem to point a hole in them. Can you double check?

ywelsch replied:
It is functionally correct, just leads to a larger DIFF than needed ;-)

bleskes replied:
that depends on you definition of correct :)

bleskes replied:
Just an idea - can we maybe decide on the diffable aspects by the previous value?

```
                T beforePart = before.get(partIter.getKey());
                if (beforePart == null) {
                    upserts.put(partIter.getKey(), partIter.getValue());
                } else if (partIter.getValue().equals(beforePart) == false) {
                    if (beforePart instanceof Diffable) {
                        diffs.put(partIter.getKey(), ((Diffable<T>)beforePart).diff(partIter.getValue()));
                    } else {
                        upserts.put(partIter.getKey(), partIter.getValue());
                    }
                }
```

If the works I think the value serializer can make be reduced to read only?

ywelsch replied:
My first implementation was like that. But then I thought about the huge amount of casts this entails. As we need the information for reading anyhow, I went with the approach above. Note that I have not made any performance tests though.

bleskes replied:
OK. Fair enough.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -19,263 +19,624 @@
-        protected final Map<String, T> adds;
+        protected final List<K> deletes;
+        protected final Map<K, Diff<T>> diffs; // incremental updates
+        protected final Map<K, T> upserts; // additions or full updates
```
this is a great way of handling the non-diffable case. nice one.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -19,263 +19,624 @@
+     * @param <K> type of map keys
+     * @param <V> type of map values
+     */
+    public static class DelegateValueSerializer<K, V> implements ValueSerializer<K, V> {
```
it seems we only use this for DiffablePrototypeValueReader, where we can inherit from DiffableValueSerializer  . Makes me wonder if we need this class...

bleskes commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -19,263 +19,624 @@
+    }
+
+    /**
+     * Implementation of ValueSerializer that serializes immutable HashSets
```
I think this should read serializes immutable Sets (hash sets is an implementation detail)

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -218,6 +229,21 @@ private IndexMetaData(String index, long version, State state, Settings settings
         this.mappings = mappings;
         this.customs = customs;
+        // fill missing slots in activeAllocationIds with empty set if needed
+        ImmutableOpenIntMap.Builder<Set<String>> filledActiveAllocationIds = null;
```
I thinks this can better go in the builder where the map isn't immutable?

ywelsch replied:
at that point, numberOfShards has not been computed yet. I would prefer moving all that precomputing code to builder, do you agree?

bleskes replied:
++

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -693,6 +754,25 @@ public Custom getCustom(String type) {
         }
 
+        public Builder putActiveAllocationIds(int shardId, Set<String> allocationIds) {
+            activeAllocationIds.put(shardId, Collections.unmodifiableSet(new HashSet(allocationIds)));
```
wondering if we should enforce immutability on this level... feels more natural to do it in the build()

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -693,6 +754,25 @@ public Custom getCustom(String type) {
+            return this;
+        }
+
+        public Builder removeActiveAllocationIds(int shardId) {
```
unused..

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -693,6 +754,25 @@ public Custom getCustom(String type) {
+            return this;
+        }
+
+        public Builder removeAllActiveAllocationIds() {
```
unused..

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -792,6 +882,21 @@ public static IndexMetaData fromXContent(XContentParser parser) throws IOExcepti
+                                        allocationIds.add(parser.text());
+                                    }
+                                }
+                                builder.putActiveAllocationIds(Integer.valueOf(shardId), Collections.unmodifiableSet(allocationIds));
```
paranoid! :) (re double immutability)

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/allocation/ActiveAllocationIdTests.java:
```diff
@@ -0,0 +1,81 @@
+        clusterState = ClusterState.builder(clusterState).routingResult(rerouteResult).build();
+
+        // active allocation ids should not be updated
+        assertThat(clusterState.getRoutingTable().shardsWithState(UNASSIGNED).size(), equalTo(3));
```
cool. Thanks!

bleskes commented on core/src/test/java/org/elasticsearch/cluster/serialization/DiffableTests.java:
```diff
@@ -96,8 +97,81 @@ public TestDiffable readFrom(StreamInput in) throws IOException {
+    }
+
+    public void testImmutableOpenIntMapDiff() throws IOException {
+        ImmutableOpenIntMap.Builder<TestDiffable> builder = ImmutableOpenIntMap.builder();
```
can we use more of randomization? create two completely random sets and try to move from one to another.. reproducible evil FTW

bleskes commented on core/src/test/java/org/elasticsearch/cluster/serialization/DiffableTests.java:
```diff
@@ -96,8 +97,81 @@ public TestDiffable readFrom(StreamInput in) throws IOException {
+    }
+
+    public void testImmutableOpenIntMapDiffOfNonDiffables() throws IOException {
+        ImmutableOpenIntMap.Builder<String> builder = ImmutableOpenIntMap.builder();
```
same.. we can maybe join the two tests into one method which we call with a parameter of whether to use diffs?

bleskes commented:
Nice one @ywelsch . Left some minor comments.

ywelsch commented:
@bleskes Another set of changes:

- Reworked DiffableTests so that it now parameterizes over the following:
  - map implementation type (JDK Map, ImmutableOpenMap, ImmutableOpenIntMap)
  - map value type (Diffable or non-diffable)
  - specific values that are added / removed. We now compare randomly-generated sets.
- Moved value extraction from IndexMetaData constructor to its builder
- `supportsDiffableValues()` is now applied after equals check
- removed unnecessary builder methods

bleskes commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -19,263 +19,596 @@
+     *
+     * Note: this implementation is ignoring the key.
+     */
+    public static class DiffablePrototypeValueReader<K, V extends Diffable<V>> extends DiffableValueSerializer<K, V> {
```
DiffablePrototypeValueReader -> DiffablePrototypeValue*Serializer* for consistency

ywelsch replied:
The prototype is only used for the "reading" part, hence the name.

bleskes replied:
yeah, got it- just found it confusing.

bleskes commented on core/src/test/java/org/elasticsearch/cluster/serialization/DiffableTests.java:
```diff
@@ -22,68 +22,324 @@
+
+            logger.debug("--> serializing diff");
+            Diff diff = calculateDiff(beforeMap, afterMap);
+            BytesStreamOutput out = new BytesStreamOutput();
```
can we only randomly use serialization? not all diffs go through it

bleskes commented on core/src/test/java/org/elasticsearch/cluster/serialization/DiffableTests.java:
```diff
@@ -22,68 +22,324 @@
+            diff.writeTo(out);
+            StreamInput in = StreamInput.wrap(out.bytes());
+            logger.debug("--> reading diff back");
+            Diff<T> diffMap = readDiffMap(in);
```
can we verify the contents of the diffs - i.e., check that the amount of upserts, removed, sub diffs are what we expect?

bleskes commented:
LGTM. Left some minor suggestions (first on the commit - sorry for the noise). I would love it if @s1monw or @imotov will also look at the changes to the diff infra as the were involved in it as well.

ywelsch commented:
pushed minor changes to address @bleskes's suggestions
- check correctness of diffmap
- randomize serialization

s1monw commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -19,263 +19,608 @@
+            return diffs;
+        }
+
+        public Map<K, T> getUpserts() {
```
can we have some javadocs what that is?

s1monw commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -19,263 +19,608 @@
+        public static final IntKeySerializer INSTANCE = new IntKeySerializer();
+
+        @Override
+        public void writeKey(Integer key, StreamOutput out) throws IOException {
```
can we throw an exception if the key is negative?

s1monw commented:
I will look again this afternoon... sorry for the delay

s1monw commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -19,263 +19,608 @@
+        /**
+         * Writes value as diff to stream if this serializer supports diffable values
+         */
+        void writeDiffTo(Diff<V> value, StreamOutput out) throws IOException;
```
I think `write` and `read` would be enough as method names? the `Diff` & `To` part is implicit from the args?

ywelsch replied:
Makes sense to throw the `To` and `From` part away. I want to keep `writeDiff` and `readDiff` though. For `readDiff`, it's necessary as it has the same method signature as `read` (except return type). To keep things symmetric for `write` and make it more verbose in the code of the `MapDiff` implementation, I'll use `writeDiff`.

s1monw commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -19,263 +19,608 @@
+        /**
+         * Computes diff if this serializer supports diffable values
+         */
+        Diff<V> computeDiff(V value, V beforePart);
```
just call this `diff(V value, V beforePart)`?

s1monw commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -19,263 +19,608 @@
+        /**
+         * Writes value to stream
+         */
+        void writeTo(V value, StreamOutput out) throws IOException;
```
also just `write`?

s1monw commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -19,263 +19,608 @@
+        /**
+         * Reads value from stream
+         */
+        V readFrom(StreamInput in, K key) throws IOException;
```
just `read`?

s1monw commented on core/src/main/java/org/elasticsearch/cluster/DiffableUtils.java:
```diff
@@ -19,263 +19,608 @@
+     *
+     * @param <K> type of map key
+     */
+    public static class ImmutableStringSetValueSerializer<K> extends NonDiffableValueSerializer<K, Set<String>> {
```
I wonder if the `Immutable` is required here, it rather confuses me though...

ywelsch replied:
removed it.

s1monw commented:
i left some cosmetic comments LGTM in general

s1monw commented:
LGTM
